File: ruby/jump_game.rb
# @param {Integer[]} nums
# @return {Boolean}
def can_jump(nums)
    can_jump_helper(nums, {})
end

def can_jump_helper(nums, cache)
    return true if nums.size <= 1

    result = (1...nums.size).any? do |i|
        new_last_index = (nums.size - 1) - i

        subset = nums[0..new_last_index]

        cache[subset] || (nums[new_last_index] >= i && can_jump_helper(subset, cache))
    end

    cache[nums] = result
    result
end

File: javascript/jump_game.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canJump = function(nums) {
    return dp(nums, nums.length, { 1: true });
};

function dp(nums, numsLength, cache) {
    if (numsLength <= 1) { return true; }

    for (var i = 0; i < numsLength - 1; i++) {
        var k = numsLength - (i + 1);

        cache[numsLength - k] = cache[numsLength - k] !== undefined ? cache[numsLength - k] : dp(nums, numsLength - k, cache);

        result = cache[numsLength - k] && nums[numsLength - 1 - k] >= k;

        if (result) { return true; }
    }

    return false;
}
